src/vector/index.ts

```diff
@@ -78,6 +78,48 @@ function checkBrowserFeatures() {
     return featureComplete;
 }
 
+/**
+ * Monkey patch some browser methods to work around a fundamental
+ * incompatability between React and any browser extension or anything
+ * else that manipulates the DOM.
+ *
+ * React issue: https://github.com/facebook/react/issues/11538
+ * Chromium issue: https://bugs.chromium.org/p/chromium/issues/detail?id=872770
+ *
+ * The workaround here is inspired by the one in a comment on the React bug
+ * (https://github.com/facebook/react/issues/11538#issuecomment-417504600)
+ * except that we catch the exception rather than sanity checking every call,
+ * and we remove the child from whatever its parent actually is rather than
+ * doing nothing (the workaround in the comment leaves the element in the DOM,
+ * which in the case of https://github.com/vector-im/riot-web/issues/13557
+ * makes it fail by adding another submit button to the page every time you
+ * try to log in, which is also a fairly terrible failure mode).
+ */
+function monkeyPatchForReact() {
+    const originalRemoveChild = Node.prototype.removeChild;
+    Node.prototype.removeChild = function(...args) {
+        try {
+            return originalRemoveChild.apply(this, args);
+        } catch (e) {
+            console.log("Caught exception from removeChild", e);
+            return originalRemoveChild.apply(args[0].parentNode, args);
+        }
+    }
+
+    const originalInsertBefore = Node.prototype.insertBefore;
+    Node.prototype.insertBefore = function(...args) {
```

Comment:
If this fails for any reason not being google translate (e.g. doing some manual DOM manipulation as we do in some places, e.g. the composer), we'll just swallow the error here.

Also, when inserting nodes in a google-translate-transformed DOM, AFAICT the node would just not get inserted? That would be quite baffling.

I'm worried of this causing inexplicable bugs down the road.

Also, if either the google translate or react implementation changes, some of these errors could surface again...

Reply:
> If this fails for any reason not being google translate (e.g. doing some manual DOM manipulation as we do in some places, e.g. the composer), we'll just swallow the error here.

True, although this would only be a problem for anything that relies on those method throwing if they're called incorrectly, which *hopefully* nothing is.

> Also, when inserting nodes in a google-translate-transformed DOM, AFAICT the node would just not get inserted? That would be quite baffling.

Yeah - I wondered whether to append the node to the start of the parent instead if it failed, so the node would be there, possibly in the wrong place. Unsure if that would be better or worse.

> I'm worried of this causing inexplicable bugs down the road, and having to think if a given problem could be `insertBefore` or `removeChild` acting up would be quite a burden IMO.

This is why I was reluctant to change their behaviour too much to insert in other places on exception, etc. Hopefully the logging should help mitigate this a bit.

> Also, if either the google translate or react implementation changes, some of these errors could surface again...

Yep, but all 3 solutions share this problem.

Reply:
> True, although this would only be a problem for anything that relies on those method throwing if they're called incorrectly, which hopefully nothing is.

It'll be easy to overlook errors like this during development if the errors are swallowed.

> Yep, but all 3 solutions share this problem.

Just disabling translation completely is the least likely to break like this I think, and IMO is not too horrible of a solution, and doesn't come with subtle pitfalls for developers.

Reply:
> > Yep, but all 3 solutions share this problem.
> 
> Just disabling translation completely is the least likely to break like this I think, and IMO is not too horrible of a solution, and doesn't come with subtle pitfalls for developers.

Yeah, it'll be fine with Google translate, it just won't prevent the error from any other browser extension etc. that manipulates the DOM.

Reply:
Fair point, although I'd personally prefer to cross that bridge when we get there (e.g. receive bug reports for a broken extension that has a reasonable use case to use with Riot).